### PR TITLE
Update Contribution Guide [Resolves #425]

### LINKED ISSUE #425: Update contribution guidelines
The contribution guidelines are quite generic and we should include more things specific to Triage, like experiment (or other object) validation, documentation, example config, dirtyduck, etc.

### PR DESCRIPTION
**Pull request by thcrock**
The contribution guide is far too generic and needed updating. This
commit, in addition to converting to Markdown, adds much detail on
testing, validation, etc.

### CODE REVIEW

**Comment by ecsalomon** [comment]
Looking through #470, I wonder if we should mention line length here.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+Report bugs at <https://github.com/dssg/triage/issues>.
+
+If you are reporting a bug, please include:
+
+-   Your operating system name and version.
+-   Any details about your local setup that might be helpful in
+    troubleshooting.
+-   Detailed steps to reproduce the bug.
+
+### Write Documentation
```
[the remark this answers is not in the document]
Comment (thcrock) [reply CONTRIBUTING.md]: Excellent point, I need to add reference for how to build/look at the docs with mkdocs

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
+Contributing
+============
+
+Contributions are welcome, and they are greatly appreciated! Every
+little bit helps, and credit will always be given.
+
+You can contribute in many ways:
+
+Types of Contributions
```
[the remark this answers is not in the document]
Comment (thcrock) [reply CONTRIBUTING.md]: Indeed, missed that

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    - [Documentation](#documentation)
+
+6.  When you're done making changes, check that your changes pass flake8
+    and the tests, including testing other Python versions with tox:
+
+        $ flake8 triage tests
+        $ python setup.py test or py.test
+        $ tox
+
+    To get flake8 and tox, just pip install them into your virtualenv.
```
[the remark this answers is not in the document]
Comment (thcrock) [reply CONTRIBUTING.md]: Hmm, I think this text may be out of date. Tox is in requirement/include/test-management and flake8 is in requirement/include/lint

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    Regardless of your Python environment, we do recommend installing `triage` as a module in development mode, via `python setup.py develop` or `pip install -e .`
+
+        $ mkvirtualenv triage
+        $ cd triage/
+        $ python setup.py develop
+
+4.  Create a branch for local development:
+
+        $ git checkout -b name-of-your-bugfix-or-feature
+
```
[the remark this answers is not in the document]
Comment (thcrock) [reply CONTRIBUTING.md]: The implication here is that the issue already exists and you are working on it.

We are combining one-time config setup (steps 1-3) with per-issue changes so maybe I should break that up to make the workflow more clear

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    The top line will be prepopulated by Github as the title of the pull request. The [Resolves #issueno] will automatically close the referenced issue when the commit is merged into master. The rest of the text will be prepopulated by Github as the content of the pull request. Ideally, you should put any useful text you want to show up in the pull request in the commit message in this way, because the commit message is available to read in the commit history long after the pull request is inconvenient to dig up. Whether or not you feel the commit is best described by a text writeup, enumerated changes, or a combination of both, the One-line summary is very useful no matter what.
+
+    To achieve this format, it's not recommended to use `git commit -m "message"`. Instead use
+
+        $ git commit
+
+    And [write the commit message in your favorite text editor](https://help.github.com/articles/associating-text-editors-with-git/)
+
+8.  Submit a pull request through the GitHub website.  If you are working directly with the DSaPP team, *assign somebody*! Pull requests should not linger be out for a long time. And tell them that you assigned them.
```
[the remark this answers is not in the document]
Comment (thcrock) [reply CONTRIBUTING.md]: I think I tried to replace one with the other and forgot to remove one. Probably just 'linger'.

[the remark this answers is not in the document]
Comment (thcrock) [reply CONTRIBUTING.md]: That's a good point. I think #2 probably makes sense.  People totally external to DSaPP aren't going to be contributing a ton of code changes, so for these very infrequent edge cases leaving it open-ended probably makes more sense than trying to call out specific people. I can add copy for this

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+
+Small Changes
+-------------
+Pull Requests take a lot of time to read. Finding defects is not the only reason to have other developers review code, but it is useful to give them a fighting chance to find defects without taking a ton of time to review. Generally, the smaller the pull request, the better. The ideal pull request is the minimal piece of code that adds value. A lot has been written on the subject; one good write-up is [The Anatomy of A Perfect Pull Request](https://medium.com/@hugooodias/the-anatomy-of-a-perfect-pull-request-567382bb6067). 
+
+
+Testing
+-------
+Triage has an extensive test suite that is run with every commit by travis-ci. Any pull request will be rejected if it breaks the travis-ci build. To help prevent yourself from encountering annoying test failures at pull request time, we strongly recommend utilizing 'Test-Driven Development' (TDD) to introduce tests along with code.
```
[the remark this answers is not in the document]
Comment (thcrock) [reply CONTRIBUTING.md]: Sure. A generic TDD document (like the wikipedia page).

The stuff I was thinking about adding would be more specific to our setup, like how we use the temporary postgres server library to accomplish DB-aware tests, and references to the testing utils (e.g. in src/tests/utils.py) that we've created to make working with specific types of objects easier.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+Many of the tests require a Postgres server, so you need the Postgres server executable installed on the system you're developing on. You don't need to create any databases, or even have the server running; the tests that need it take care of starting a server, creating databases, and tearing it all down when they're done.
```
[the remark this answers is not in the document]
Comment (thcrock) [reply CONTRIBUTING.md]: Good point, we encapsule that in https://github.com/dssg/triage/blob/master/requirement/test.txt but I should point that out

**Comment by thcrock** [comment]
Line length is part of the PEP8, I'm not sure if it's worth elevating above other PEP8 concepts.

**Pull request merged**